velox/functions/prestosql/JsonFunctions.cpp

```diff
@@ -130,45 +251,103 @@ class JsonParseFunction : public exec::VectorFunction {
   }
 
  private:
-  simdjson::error_code parse(size_t size) const {
+  simdjson::error_code parse(size_t size, JsonViews& jsonViews) const {
     simdjson::padded_string_view paddedInput(
         paddedInput_.data(), size, paddedInput_.size());
     SIMDJSON_ASSIGN_OR_RAISE(auto doc, simdjsonParse(paddedInput));
-    SIMDJSON_TRY(validate<simdjson::ondemand::document&>(doc));
+    SIMDJSON_TRY(validate<simdjson::ondemand::document&>(doc, jsonViews));
     if (!doc.at_end()) {
       return simdjson::TRAILING_CONTENT;
     }
     return simdjson::SUCCESS;
   }
 
   template <typename T>
-  static simdjson::error_code validate(T value) {
+  static simdjson::error_code validate(T value, JsonViews& jsonViews) {
     SIMDJSON_ASSIGN_OR_RAISE(auto type, value.type());
     switch (type) {
       case simdjson::ondemand::json_type::array: {
         SIMDJSON_ASSIGN_OR_RAISE(auto array, value.get_array());
+
+        jsonViews.push_back(kArrayStart);
+        auto jsonViewsSize = jsonViews.size();
         for (auto elementOrError : array) {
           SIMDJSON_ASSIGN_OR_RAISE(auto element, elementOrError);
-          SIMDJSON_TRY(validate(element));
+          SIMDJSON_TRY(validate(element, jsonViews));
+          jsonViews.push_back(kSeparator);
+        }
+
+        // If the array is not empty, remove the last separator.
+        if (jsonViews.size() > jsonViewsSize) {
+          jsonViews.pop_back();
         }
+
+        jsonViews.push_back(kArrayEnd);
+
         return simdjson::SUCCESS;
       }
+
       case simdjson::ondemand::json_type::object: {
         SIMDJSON_ASSIGN_OR_RAISE(auto object, value.get_object());
+
+        std::vector<std::pair<std::string_view, JsonViews>> objFields;
```

Comment:
It's suboptimal that we have to allocate this complex thing at every object node, but we can leave this later until we need to optimize this.

This is a typical example of why C++ sometimes is not the right interface for high performance programming and lures people into something inefficient. If we had programed it in C, we would likely have chosen a more efficient data structure.